python/src/equistore/operations/slice.py

```diff
@@ -0,0 +1,395 @@
+                + "' - doesn't exist in input `TensorBlock`."
+            )
+    if properties_to_slice is not None:
+        assert isinstance(
+            properties_to_slice, Labels
+        ), "properties_to_slice must be passed as a equistore.Labels object."
+        property_names = (
+            tensorblock.property_names
+            if isinstance(tensorblock, TensorMap)
+            else tensorblock.properties.names
+        )
+        for name in properties_to_slice.names:
+            assert name in property_names, (
+                "Invalid property name '"
+                + name
+                + "' - doesn't exist in input `TensorBlock`."
+            )
+
+    # Slice TensorMap and issue warning if the output block is empty
+    sliced_tensorblock = _slice_tensorblock(
+        tensorblock,
+        samples_to_slice=samples_to_slice,
+        properties_to_slice=properties_to_slice,
+    )
+    if np.any(np.array(sliced_tensorblock.values.shape) == 0):
+        warnings.warn(
+            "Your input TensorBlock is now empty, based on your choice of samples "
+            + "and/or properties to slice by. "
+        )
+
+    return sliced_tensorblock
+
+
+def _slice_tensorblock(
+    tensorblock: TensorBlock, samples_to_slice=None, properties_to_slice=None
+) -> TensorBlock:
+    """
+    Slices an input :py:class:`TensorBlock` along the samples and/or properties
+    dimension(s). `samples_to_slice` and `properties_to_slice` are
+    :py:class:`Labels` objects that specify the samples/properties
+    (respectively) names and indices that should be sliced, i.e. kept in the
+    output :py:class:`TensorBlock`.
+
+    Note that either `samples_to_slice` or `properties_to_slice`, or both,
+    should be specified as input.
+
+    :param tensorblock: the input :py:class:`TensorBlock` to be sliced.
+    :param samples_to_slice: a :py:class:`Labels` object containing the names
+        and indices of samples to keep in the sliced output
+        :py:class:`TensorBlock`. Default value of None indicates no slicing
+        along the samples dimension should occur.
+    :param properties_to_slice: a :py:class:`Labels` object containing the names
+        and indices of properties to keep in the sliced output
+        :py:class:`TensorBlock`. Default value of None indicates no slicing
+        along the properties dimension should occur.
+
+    :return new_block: a :py:class:`TensorBlock` that corresponds to the sliced
+        input.
+    """
+
+    # Generate arrays of bools indicating which samples indices to keep upon slicing.
+    if samples_to_slice is None:
+        # No samples slicing wanted; keep all properties indices.
+        samples_filter = np.array([True] * len(tensorblock.samples))
+    else:
+        samples_filter = np.array(
+            [
+                sample in samples_to_slice.tolist()
+                for sample in tensorblock.samples[list(samples_to_slice.names)].tolist()
+            ]
```

Comment:
I'm wondering how expensive this and the other list comprehension are going to get. The main issue I have is that this is going to be O(n^2) (`sample in ...` does a full loop search), so it might be better to use specialized data structures such as a set (converting `samples_to_slice.tolist()` to a set)

Reply:
Intervening just to say that indeed it'd be good to make sure that these utility functions are not going to become unexpected computational bottlenecks further down the line.

Reply:
Understood, have talked to Guillaume and will have a go at some line profiling and optimisation.